src/client.rs

```diff
@@ -138,6 +138,11 @@ fn validate_sequence_set(
 #[derive(Debug)]
 pub struct Session<T: Read + Write> {
     conn: Connection<T>,
+
+    // Capabilities are almost guaranteed to chance if encryption state or authentication state
+    // changes, so caching them in `Connection` is inappropiate.
+    capability_cache: Option<Capabilities>,
```

Comment:
I don't know that I believe that we shouldn't have the cache be in `Connection`. We can always just forcibly-reset it whenever the relevant underlying state changes (e.g., in `login`). What do you think?

Reply:
I had two arguments for not caching the capabilities in the `Connection`: Firstly and definitely the weaker of the two: Separation of concerns, namely doing so makes `Connection` more than just a very thin shim over the underlying `io::{Read + Write}`. Secondly and my main argument is that caching in `Connection` has a smaller fuck-up-protection factor for the maintainers, i.e. "*us, three years older*". Clearing the cache becomes an active action that needs to be thought of, and *forgetting* to do so is still syntactically and semantically valid code, just one

Reply:
While I think that's true, I also think the current split is a foot-gun because it means we have to replicate the cache state and logic in two places. If it was on `Connection`, it would be in one place, which means one place to make improvements / bugfixes over time. I think I'd err on the side of simplicity here, as it'll be quite rare that we add a new way to wrap `Connection`. And if we do, we're likely to do so by copy-pasting an existing wrapper first.